packages/babel-plugin-react-i18n/src/BabelPluginReactI18n.ts

```diff
@@ -0,0 +1,97 @@
+import {TemplateBuilder} from '@babel/template';
+import * as Types from '@babel/types';
+import {NodePath} from '@babel/traverse';
+import stringHash from 'string-hash';
+import path from 'path';
+import fs from 'fs';
+
+interface State {
+  importer(importArg: Types.ImportDeclaration): void;
+}
+
+const WITH_I18N_DECORATOR_NAME = 'withI18n';
+
+export default function injectWithI18nArguments({
+  types: t,
+  template,
+}: {
+  types: typeof Types;
+  template: TemplateBuilder<Types.ImportDeclaration | Types.ObjectExpression>;
+}) {
+  const fallbackTranslationsImport = template(
+    `import enTranslations from './translations/en.json';`,
+    {sourceType: 'module'},
+  );
+  function i18nParams({ID}) {
+    return template(
+      `withI18n({
+      id: '${ID}',
+      fallback: enTranslations,
+      async translations(locale) {
+          try {
+            const dictionary = await import(/* webpackChunkName: '${ID}-i18n' */ \`./translations/$\{locale}.json\`);
+            return dictionary;
+          } catch (err) {}
+        },
+    })`,
+      {
+        sourceType: 'module',
+        plugins: ['dynamicImport'],
+        preserveComments: true,
+      },
+    )();
+  }
+
+  return {
+    visitor: {
+      Program(_path: NodePath<Types.Program>, state: State) {
+        const lastImport = _path
+          .get('body')
+          .filter(subPath => subPath.isImportDeclaration())
+          .pop();
+        if (lastImport) {
+          state.importer = importDecl => lastImport.insertAfter(importDecl);
+        } else {
+          state.importer = importDecl =>
+            _path.get('body')[0].insertBefore(importDecl);
+        }
+      },
+      CallExpression(_path: NodePath<Types.CallExpression>, state: State) {
+        const expr = _path.node;
+        const {callee} = expr;
+        if (
+          t.isIdentifier(callee) &&
+          (callee as Types.Identifier).name === WITH_I18N_DECORATOR_NAME &&
+          expr.arguments.length === 0
+        ) {
+          const {filename, filenameRelative} = this.file.opts;
+          const enJSONPath = path.resolve(
+            path.dirname(filename),
+            'translations',
+            'en.json',
+          );
+          if (!fs.existsSync(enJSONPath)) {
+            return;
+          }
+          state.importer(
+            fallbackTranslationsImport() as Types.ImportDeclaration,
```

Comment:
To me this logic feels backwards. I’d rather do the following:

1. Check if the file actually includes an import for React-i18n
2. Check if that binding is ever used without arguments
3. Check if there is a translation file
4. Process each no-argument i18n call if all of the above are true

Reply:
that is technically more correct, I suppose. I considered that approach, but I was worried about the second step. I suppose we could use `state` to do this entire thing in a single traversal. 🤔